src/odemis/acq/fastem.py

```diff
@@ -18,73 +20,398 @@
+        self._roc = roa.roc  # region of calibration object
+        self._path = path  # sub-directories on external storage
+        self._future = future
+
+        # Dictionary containing the single field images with index as key: e.g. {(0,1): DataArray}.
+        self.megafield = {}
+        self.field_idx = (0, 0)
+
+        # TODO the .dataContent might need to be set somewhere else in future when using a live stream for
+        #  display of thumbnail images -> .dataContent = "thumbnail"
+        # set size of returned data array
+        # The full image data is directly stored via the asm on the external storage.
+        self._detector.dataContent.value = "empty"  # dataArray of shape (0,0) is returned with some MD
+
+        # list of field image indices that still need to be acquired {(0,0), (1,0), (0,1), ...}
+        self._fields_remaining = set(self._roa.field_indices)  # Used for progress update.
+
+        # keep track if future was cancelled or not
+        self._cancelled = False
+
+        # Threading event, which keeps track of when image data has been received from the detector.
+        self._data_received = threading.Event()
+
+    def run(self):
+        """
+        Runs the acquisition of one ROA (megafield).
+        :returns:
+            megafield: (list of DataArrays) A list of the raw image data. Each data array (entire field, thumbnail,
+                or zero array) represents one single field image within the roa (megafield).
+            exception: (Exception or None) Exception raised during the acquisition. If some single field image data has
+                already been acquired, exceptions are not raised, but returned.
+        :raise:
+            Exception: If it failed before any single field images were acquired or if acquisition was cancelled.
+        """
+
+        # set the sub-directories (<acquisition date>/<project name>) and megafield id
+        self._detector.filename.value = os.path.join(self._path, self._roa.name.value)
+
+        exceptions = None
+
+        # Get the estimated time for the roa.
+        total_roa_time = self._roa.estimate_roa_time()
+        # No need to set the start time of the future: it's automatically done when setting its state to running.
+        self._future.set_progress(end=time.time() + total_roa_time)  # provide end time to future
+        logging.info("Starting acquisition of mega field, with expected duration of %f s", total_roa_time)
+
+        dataflow = self._detector.data
+
+        try:
+            logging.debug("Starting megafield acquisition.")
+            dataflow.subscribe(self.image_received)
+
+            # Acquire the single field images.
+            self.acquire_roa(dataflow)
+
+            # In case the acquisition was cancelled, before the future returned, raise cancellation error
+            # after unsubscribing from the dataflow
+            if self._cancelled:
+                self._future.cancel()
```

Comment:
This will call (again) the `cancel()` function. It shouldn't be needed. Just drop this line.